ipld/read.go

```diff
@@ -18,190 +17,165 @@ import (
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
+	if err != nil {
+		return nil, err
 	}
-	return out
-}
 
-type index struct {
-	row uint32
-	col uint32
+	return extended, batchAdder.Commit()
 }
 
-type indexedShare struct {
-	data []byte
-	index
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	to            int
+	rootCids      []cid.Cid
 }
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+func pickRandomQuadrant(roots [][]byte) (*quadrant, error) {
+	edsWidth := len(roots)
+	// get the random number from 1 to 4.
+	q := rand.Intn(4) + 1 //nolint:gosec
+	quadrant := &quadrant{rootCids: make([]cid.Cid, edsWidth/2)}
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/*
+		| 1 | 2 |
+		| 3 | 4 |
+	*/
+	// choose subtree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
 
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
-}
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		quadrant.to = edsWidth
+	} else {
+		quadrant.to = edsWidth / 2
+	}
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+	var err error
+	for index, counter := quadrant.from, 0; index < quadrant.to; index++ {
+		quadrant.rootCids[counter], err = plugin.CidFromNamespacedSha256(roots[index])
+		if err != nil {
+			return nil, err
+		}
+		counter++
+	}
+	return quadrant, nil
+}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	dataSquare [][]byte,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < len(roots.rootCids); i++ {
+		i := i
+		errGroup.Go(func() error {
+			leaves, err := GetSubtreeLeaves(ctx, roots.rootCids[i], dag, roots.isLeftSubtree, uint32(len(roots.rootCids)/2))
+			if err != nil {
+				return err
+			}
+			length := len(roots.rootCids)
+			for leafIdx, leaf := range leaves {
+				// shares from the right subtree should be
+				// inserted after all shares from the left subtree
+				if !roots.isLeftSubtree {
+					leafIdx += len(leaves)
+				}
+				shareData := leaf.RawData()[1:]
+				dataSquare[((i+roots.from)*length*2)+leafIdx] = shareData[NamespaceSize:]
```

Comment:
Can you just document here how you calculate the position/index of where to add the share data inside the `dataSquare` ? Something like: 

```suggestion
                                 // position is calculated by < insert explanation>
                                 position := ((i+roots.from)*length*2)+leafIdx
				dataSquare[position] = shareData[NamespaceSize:]
```